<font size="+3"><b>Assignment 2: Linear Models and Validation Metrics</b></font>

***
* **Full Name** = Sudarshan Naicker 
* **UCID** = 30162797
***

<font color='Blue'>In this assignment, you will need to write code that uses linear models to perform classification and regression tasks. You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.</font>

You can use the Table of Content on the left side of this notebook to efficiently navigate within this documents.

|                **Question**                | **Point** |
|:------------------------------------------:|:---------:|
|         **Part 1: Classification**         |           |
|          Step 0: Import Libraries          |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    1.5    |
| Step 3: Implement Machine Learning   Model |           |
|           Step 4: Validate Model           |           |
|          Step 5: Visualize Results         |     4     |
|                  Questions                 |     4     |
|             Process Description            |     4     |
|           **Part 2: Regression**           |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    0.5    |
| Step 3: Implement Machine Learning   Model |     1     |
|            Step 4: Validate Mode           |     1     |
|          Step 5: Visualize Results         |     1     |
|                  Questions                 |     2     |
|             Process Description            |     4     |
|  **Part 3:   Observations/Interpretation** |   **3**   |
|           **Part 4: Reflection**           |   **2**   |
|                  **Total**                 |   **30**  |
|                                            |           |
|                  **Bonus**                 |           |
|         **Part 5: Bonus Question**         |   **4**   |

# **Part 1: Classification (14.5 marks total)**

|                **Question**                | **Point** |
|:------------------------------------------:|:---------:|
|         **Part 1: Classification**         |           |
|          Step 0: Import Libraries          |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    1.5    |
| Step 3: Implement Machine Learning   Model |           |
|           Step 4: Validate Model           |           |
|          Step 5: Visualize Results         |     4     |
|                  Questions                 |     4     |
|             Process Description            |     4     |
|                  **Total**                 |  **14.5** |

You have been asked to develop code that can help the user determine if the email they have received is spam or not. Following the machine learning workflow described in class, write the relevant code in each of the steps below:


## **Step 0:** Import Libraries

In [2]:
import numpy as np
import pandas as pd

## **Step 1:** Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

Use the yellowbrick function `load_spam()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [3]:
# TO DO: Import spam dataset from yellowbrick library
from yellowbrick.datasets import load_spam

# TO DO: Print size and type of X and y
X, y = load_spam()
print('Size of X:', X.size)
print('Type of X:', type(X))
print('Size of y:', y.size)
print('Type of y:', type(y))


Size of X: 262200
Type of X: <class 'pandas.core.frame.DataFrame'>
Size of y: 4600
Type of y: <class 'pandas.core.series.Series'>


## **Step 2:** Data Processing (1.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [4]:
# TO DO: Check if there are any missing values and fill them in if necessary

missing_values_X = X.isnull().sum()
missing_values_y = y.isnull().sum()
print("Missing values in X:\n", missing_values_X)
print("\nMissing values in y: ", missing_values_y)

#No missing Values

Missing values in X:
 word_freq_make                0
word_freq_address             0
word_freq_all                 0
word_freq_3d                  0
word_freq_our                 0
word_freq_over                0
word_freq_remove              0
word_freq_internet            0
word_freq_order               0
word_freq_mail                0
word_freq_receive             0
word_freq_will                0
word_freq_people              0
word_freq_report              0
word_freq_addresses           0
word_freq_free                0
word_freq_business            0
word_freq_email               0
word_freq_you                 0
word_freq_credit              0
word_freq_your                0
word_freq_font                0
word_freq_000                 0
word_freq_money               0
word_freq_hp                  0
word_freq_hpl                 0
word_freq_george              0
word_freq_650                 0
word_freq_lab                 0
word_freq_labs                0
word_freq_telnet  

For this task, we want to test if the linear model would still work if we used less data. Use the `train_test_split` function from sklearn to create a new feature matrix named `X_small` and a new target vector named `y_small` that contain **5%** of the data.

In [15]:
# TO DO: Create X_small and y_small
from sklearn.model_selection import train_test_split

X_small, _, y_small, _ = train_test_split(X, y, test_size=0.95, random_state=42)

print(y_small.shape)

(230,)


## **Step 3:** Implement Machine Learning Model

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with three different datasets:
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`

## **Step 4:** Validate Model

Calculate the training and validation accuracy for the three different tests implemented in Step 3

## **Step 5:** Visualize Results (4 marks for steps 3-5)

1. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [6]:
# TO DO: ADD YOUR CODE HERE FOR STEPS 3-5
# Note: for any random state parameters, you can use random_state = 0
# HINT: USING A LOOP TO STORE THE DATA IN YOUR RESULTS DATAFRAME WILL BE MORE EFFICIENT
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
import pandas as pd

# Initializing the logistic regression model 
model = LogisticRegression(max_iter=2000)

# Function to train and validate the model
def model_trainer(model, X, y):
    X_train, X_validation, y_train, y_validation = train_test_split(X, y, test_size=0.2, random_state=0)
    
    # Training the model
    model.fit(X_train, y_train)
    
    # Calculating the training and validation accuracies
    training_accuracy = accuracy_score(y_train, model.predict(X_train))
    validation_accuracy = accuracy_score(y_validation, model.predict(X_validation))
    
    return X.size, training_accuracy, validation_accuracy


# Creating the results DataFrame
results = pd.DataFrame(columns=['Data size', 'Training accuracy', 'Validation accuracy'])

# Implementing the machine learning model with three different datasets
datasets = {
    'X and y Dataset': (X, y),
    'First two columns': (X.iloc[:, :2], y),
    'X_small & y_small Dataset': (X_small, y_small)
}

# training each dataset and storing the scores in results df
for name, (data, target) in datasets.items():
    results.loc[name] =  model_trainer(model, data, target)


results

Data size  Training accuracy  Validation accuracy
X and y Dataset             262200.0           0.927989             0.936957
First two columns             9200.0           0.614946             0.593478
X_small & y_small Dataset    13110.0           0.961957             0.891304

## **Questions (4 marks)**
1. How do the training and validation accuracy change depending on the amount of data used? Explain with values.
2. In this case, what do a false positive and a false negative represent? Which one is worse?

<font color='Green'><b>YOUR ANSWERS HERE</b></font>

1) Different sizes of data changes the scores drasticaly:
    - X and y Dataset: since we have used all 57 Features with 4600 data for each feature which is a more than enough for training and      
      validation. Hence we see a score of 92.7% for training accuracy and 93.6% for  validation accuracy.
    - First two columns: we have reduced the dataset to only 2 features hence the accuracies drops significantly, as seen the training accuracy score is only 61.4 % and the Validation accuracy is only 59.3 %  This is because the model has been trained on smaller dataset which is not sufficient for the model to learn the relationships between the spam and non-spam emails.
    - small_X and small_y: though we only use 5% for the training dataset we see a high score on both traning and validation of 96% and 89% respectively this shows that a minimum amount of dataset is good enough for the model to learn the complex patterns. However, more the training and testing set available more the information will be received by any machine learning model and better it will perform as seen in results the validation score has been decreased to 89% which is low compared to the full dataset. 

2) In the case of the email's dataset, 
  - Flase positive (FP) means, if a model predicts a spam email it is more likely to predict an non-spam email as spam too (Type I error).
  - Flase negative (FN) means, if a model predicts a correct emails then it is more likely to predict the spam emails as non-spam (Type II error).

  - In this case both FP and FN are bad, but i believe that FP is more worse than FN cause having a spam email classified as non-spam could lead to serious issues like having sensitve information hacked if the user uses the email etc.,

## **Process Description (4 marks)**
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

<font color='Green'><b>DESCRIBE YOUR PROCESS HERE</b></font>

1) I have used the ensf 444 labs, mainly the linear_regression_lab2 was very helful to for the train_test_split() and model_fit(), to train the model. I had to review all 3 labs for the basics like shape/size of datset, accuracy_score() function etc.
2) I completed Part 1 in ascending order from step 1-5.
4) I had challenges with creating the results dataframe to print the value of the data size training and validation scores as I was calling the model_trainer() funtion each time for the different dataset but after reading the HINT to use 'For loop' I imporvised and created a for loop to call the function this way it is easier to add more datasets in the future.

# **Part 2: Regression (10.5 marks total)**

| **Question**                               | **Point** |
|--------------------------------------------|-----------|
| **Part 2: Regression**                     |           |
| Step 1: Data Input                         | 1         |
| Step 2: Data Processing                    | 0.5       |
| Step 3: Implement Machine Learning   Model | 1         |
| Step 4: Validate Mode                      | 1         |
| Step 5: Visualize Results                  | 1         |
| Questions                                  | 2         |
| Process Description                        | 4         |
| **Total**                                  | **10.5**  |

For this section, we will be evaluating concrete compressive strength of different concrete samples, based on age and ingredients. You will need to repeat the steps 1-4 from Part 1 for this analysis.

## **Step 1:** Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

Use the yellowbrick function `load_concrete()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [7]:
# TO DO: Import spam dataset from yellowbrick library
from yellowbrick.datasets import load_concrete

# TO DO: Print size and type of X and y
X2, y2 = load_concrete()
print('Size of X:', X2.size)
print('Type of X:', type(X2))
print('Size of y:', y2.size)
print('Type of y:', type(y2))

Size of X: 8240
Type of X: <class 'pandas.core.frame.DataFrame'>
Size of y: 1030
Type of y: <class 'pandas.core.series.Series'>


## **Step 2:** Data Processing (0.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [8]:
# TO DO: Check if there are any missing values and fill them in if necessary
missing_values_X = X2.isnull().sum()
missing_values_y = y2.isnull().sum()
print("Missing values in X:\n", missing_values_X)
print("\nMissing values in y: ", missing_values_y)

Missing values in X:
 cement    0
slag      0
ash       0
water     0
splast    0
coarse    0
fine      0
age       0
dtype: int64

Missing values in y:  0


## **Step 3:** Implement Machine Learning Model (1 mark)

1. Import `LinearRegression` from sklearn
2. Instantiate model `LinearRegression()`.
3. Implement the machine learning model with `X` and `y`

In [9]:
from sklearn.linear_model import LinearRegression

In [10]:
# TO DO: ADD YOUR CODE HERE
# Note: for any random state parameters, you can use random_state = 0

model2 = LinearRegression()
X_train_2, X_validation_2, y_train_2, y_validation_2 = train_test_split(X2, y2, test_size=0.2, random_state=0)
model2.fit(X_train_2, y_train_2)

LinearRegression()

## **Step 4:** Validate Model (1 mark)

Calculate the training and validation accuracy using mean squared error and R2 score.

In [11]:
# TO DO: ADD YOUR CODE HERE
# Calculate Mean Squared Error (MSE) for both sets
from sklearn.metrics import mean_squared_error, r2_score

training_prediction = model2.predict(X_train_2)
validation_prediction = model2.predict(X_validation_2)

mse_training = mean_squared_error(y_train_2, training_prediction)
mse_validation = mean_squared_error(y_validation_2, validation_prediction)

# Calculate R² score for both sets
r2_training = r2_score(y_train_2, training_prediction)
r2_validation = r2_score(y_validation_2, validation_prediction)


## **Step 5:** Visualize Results (1 mark)
1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy, and index: MSE and R2 score
2. Add the accuracy results to the `results` DataFrame
3. Print `results`

In [16]:
# TO DO: ADD YOUR CODE HERE

results2 = pd.DataFrame(index=['MSE', 'R2 score'], columns=['Training accuracy', 'Validation accuracy'])


results2.loc['MSE', 'Training accuracy'] = mse_training
results2.loc['MSE', 'Validation accuracy'] = mse_validation
results2.loc['R2 score', 'Training accuracy'] = r2_training
results2.loc['R2 score', 'Validation accuracy'] = r2_validation

results2

Training accuracy Validation accuracy
MSE             110.345501           95.635335
R2 score          0.609071            0.636898

## **Questions (2 marks)**
1. Did using a linear model produce good results for this dataset? Why or why not?

We can see the R2 scores are around 60% which indicates that the linear model has a moderate fit of the dataset. This means that the linear model is not able to handle all the datapoints i.e., has captured some of the variance in the target variable, but not all. If we were aiming for a high precision then the linear model's performance is not optimal.

## **Process Description (4 marks)**
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

<font color='Green'><b>Explain YOUR PROCESS here:</b></font>


1) I took the help of the linear_regression_lab_2 as it had a linear_regression() example with code for calculating MSE.
2) I completed in an ascending order from steps 1-5.
3) No, I didn't need to use generative AI
4) No challenges as this was very simillar to part 1: Classifications hence when i started this section it was much easier as i was already 'trained' (see what I did there?)

# **Part 3: Observations/Interpretation (3 marks)**

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


<font color='Green'><b>ADD YOUR FINDINGS HERE</b></font>

Part 1:
As per the results we can conclude that higher the dataset size, better with the acuraccy scores, as you can see the validation scores for only first 2 columns, whose size was 9200, was the least i.e., 61% only. When we increased the size of the dataset the validation scores increased, 89% for size 13,110 and 93.7% for the 262,200 dataset size. also an interesting point to note here is that even though we are using only 5% of the dataset (13110), the training score is 96% which is higher than the training score for the full dataset (262,220) which is 92% only, this could be due to many reason, but what best reasonates with me is that the 5% dataset is large enough for the model to learn the patterns but small enough to memorize the relationships and have a perfect score. 

Part 2:
we can see that the R2 scores doesnt drastically change for the training and validation test this shows that the model is able to accomdate a moderate amount of variance even after being trained this might be because there are some datapoints that are outside the linear fit.

The MSE scores do change from 110 to 95, this means that the model is able to make better predictions for validation set than the training set since lower the MSE score the closer is the fit of the model to the data. 

# **Part 4: Reflection (2 marks)**
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


<font color='Green'><b>ADD YOUR THOUGHTS HERE</b></font>

I enjoyed learning how to train the model on the dataset and calculating the R2, MSE and accuracy scores. I didnt find this assignment very challenging but I had to research a bit to understand how training and validation works, the labs and the lecture notes helped me understand the code and definition respectively. 

# **Part 5: Bonus Question (4 marks)**

Repeat Part 2 with Ridge and Lasso regression to see if you can improve the accuracy results. Which method and what value of alpha gave you the best R^2 score? Is this score "good enough"? Explain why or why not.

**Remember**: Only test values of alpha from 0.001 to 100 along the logorithmic scale.

In [13]:
# TO DO: ADD YOUR CODE HERE

<font color='Green'><b>ADD YOUR ANSWER HERE</b></font>